gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisher.java

```diff
@@ -299,14 +270,75 @@ private ParallelRunner getParallelRunner(FileSystem fs) {
     return this.parallelRunners.get(uri);
   }
 
-  @Override
-  public void publishMetadata(Collection<? extends WorkUnitState> states) throws IOException {
-    // Nothing to do
+  private WorkUnitPublishGroup getWorkUnitPublishGroup(WorkUnitState state) throws IOException {
+    for (int branchId = 0; branchId < this.numBranches; branchId++) {
+      // The directory where the workUnitState wrote its output data.
+      Path writerOutputDir = WriterUtils.getWriterOutputDir(state, this.numBranches, branchId);
+
+      if (!this.writerFileSystemByBranches.get(branchId).exists(writerOutputDir)) {
+        LOG.warn(String.format("Branch %d of WorkUnit %s produced no data", branchId, state.getId()));
+        continue;
+      }
+
+      // The directory where the final output directory for this job will be placed.
+      // It is a combination of DATA_PUBLISHER_FINAL_DIR and WRITER_FILE_PATH.
+      Path publisherOutputDir = getPublisherOutputDir(state, branchId);
+
+      return new WorkUnitPublishGroup(branchId, writerOutputDir, publisherOutputDir);
```

Comment:
No, it should head down all branches.  From my understanding, each writerOutputDir is in one and only one branch.  If that is true, this code will return when the first branch that contains the writerOutputDir is found.